src/finalfusion/subword/hash_indexers.pyx

```diff
@@ -0,0 +1,325 @@
+    FastTextIndexer
+
+    FastTextIndexer is a hash-based subword indexer. It hashes n-grams with (a slightly faulty)
+    FNV-1a variant and maps the hash to a predetermined bucket space.
+
+    N-grams can be indexed directly through the `__call__` method or all n-grams in a string
+    can be indexed in bulk through the `subword_indices` method.
+    """
+    cdef uint32_t _min_n
+    cdef uint32_t _max_n
+    cdef uint64_t _n_buckets
+
+    def __init__(self, n_buckets=2000000, min_n=3, max_n=6):
+        assert max_n >= min_n > 0, \
+            f"Min_n ({min_n}) must be greater than 0, max_n ({min_n}) must be >= min_n"
+        assert 0 < n_buckets <= UINT32_MAX, \
+            f"n_buckets ({n_buckets}) needs to be between 0 and {pow(2, 32)}"
+        self._min_n = min_n
+        self._max_n = max_n
+        self._n_buckets = n_buckets
+
+    def __call__(self, str ngram):
+        cdef bytes b_ngram = ngram.encode("utf8")
+        return ft_hash_ngram(b_ngram, 0, len(b_ngram)) % self._n_buckets
+
+    @property
+    def n_buckets(self):
+        """
+        Get the number of buckets.
+
+        Returns
+        -------
+        n_buckets : int
+            Number of buckets
+        """
+        return self._n_buckets
+
+    @property
+    def min_n(self):
+        """
+        The lower bound of the n-gram range.
+
+        Returns
+        -------
+        min_n : int
+            Lower bound of n-gram range
+        """
+        return self._min_n
+
+    @property
+    def max_n(self):
+        """
+        The upper bound of the n-gram range.
+
+        Returns
+        -------
+        max_n : int
+            Upper bound of n-gram range
+        """
+        return self._max_n
+
+    @cython.cdivision(True)
+    cpdef subword_indices(self,
+                          str word,
+                          uint64_t offset=0,
+                          bracket=True,
+                          with_ngrams=False):
+        """
+        Get the subword indices for a word.
+        
+        Parameters
+        ----------
+        word : str
+            The string to extract n-grams from
+        offset : int
+            The offset to add to the index, e.g. the length of the word-vocabulary.
+        bracket : bool
+            Toggles bracketing the input string with `<` and `>`
+        with_ngrams : bool
+            Toggles returning tuples of (ngram, idx)
+        
+        Returns
+        -------
+        indices : list
+            List of n-gram indices, optionally as `(str, int)` tuples.
+        
+        Raises
+        ------
+        TypeError
+            If `word` is None.
+        """
+        if word is None:
+            raise TypeError("Can't extract ngrams for None")
+        cdef unsigned int start, end
+        cdef Py_ssize_t i, j
+        cdef uint64_t h
+        if bracket:
+            word = "<%s>" % word
+        cdef bytes b_word = word.encode("utf-8")
+        cdef const unsigned char*b_word_ptr = b_word
```

Comment:
space between `char` and `*`

Reply:
Just another comment: `char` may not be a valid type here, depending on what `word.encode` returns. `char` is only defined by the C standard to be **at least** 1 byte (though POSIX refines it). If you want to be absolutely correct, you probably want to use `uint8_t` here in place of `unsigned char`.